File: 01-1000/1-50/01-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    const map = {};
    for (let i = 0; i < nums.length; i++) {
        const n = nums[i];

        if (map[n] !== undefined) {
            return [i, map[n]];
        } else {
            map[target - n] = i;
        }
    }
};

File: 01-1000/01-50/01-two-sum.rs
use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map = HashMap::new();
        for (idx, num) in nums.into_iter().enumerate() {
            let idx = idx as i32; // fuck you leetcode
            if let Some(&other_idx) = map.get(&num) {
                return vec![idx, other_idx];
            }
            let _ = map.insert(target - num, idx);
        }
        unreachable!();
    }
}
